app/services/tool_filter_service.rb

```diff
@@ -0,0 +1,102 @@
+        {
+          tool_group_suggestions_weight: tool_group.suggestions_weight,
+          resource_id: resource.resource_id,
+          resource_tool_group_suggestions_weight: resource.suggestions_weight
+        }
+      end
+    end
+
+    result.group_by { |hash| hash[:resource_id] }
+  end
+
+  def match_params(tool_group, params)
+    country = params["filter"]["country"]&.upcase
+    languages = params["filter"]["languages"]
+    openness = params["filter"]["openness"].to_i
+    confidence = params["filter"]["confidence"].to_i
+
+    return true if no_rules_for(tool_group)
+    return true if language_rule(tool_group, languages)
+
+    # Rule Countries
+    if tool_group.rule_countries.any?
+      if country
+        country_positive_match = tool_group.rule_countries.any? { |o| o.countries.include?(country) && !o.negative_rule }
+        country_negative_match = tool_group.rule_countries.any? { |o| o.countries.include?(country) && o.negative_rule }
+        return false if !country_positive_match || country_negative_match
+      elsif tool_group.rule_countries.any?(&:negative_rule)
+        return false
+      end
+    end
+
+    # Rule Languages
+    if tool_group.rule_languages.any?
+      negative_rule = tool_group.rule_languages.any?(&:negative_rule)
+      language_positive_match = tool_group.rule_languages.any? { |o| (languages - o.languages).empty? && !o.negative_rule }
+      language_negative_match = tool_group.rule_languages.any? { |o| !(languages - o.languages).empty? && o.negative_rule }
+      return false unless negative_rule ? language_negative_match : language_positive_match
+    end
+
+    # Rule Praxes
+    if tool_group.rule_praxes.any?
+      # Openness
+      openness_positive_match = tool_group.rule_praxes.any? { |o| o.openness.include?(openness) && !o.negative_rule }
+      openness_negative_match = tool_group.rule_praxes.any? { |o| o.openness.include?(openness) && o.negative_rule }
+      return false if !openness_positive_match || openness_negative_match
+
+      # Confidence
+      confidence_positive_match = tool_group.rule_praxes.any? { |o| o.confidence.include?(confidence) && !o.negative_rule }
+      confidence_negative_match = tool_group.rule_praxes.any? { |o| o.confidence.include?(confidence) && o.negative_rule }
+      return false if !confidence_positive_match || confidence_negative_match
+    end
+
+    true
+  end
+
+  def no_rules_for(tool_group)
+    tool_group.rule_languages.none? &&
+      tool_group.rule_praxes.none? &&
+      tool_group.rule_countries.none?
+  end
+
+  # Returns true if the tool group has a single rule with a language/s
+  # present in the 'languages' array, otherwise returns false.
+  def language_rule(tool_group, languages)
+    return false unless tool_group.rule_languages.one?
+    return false unless tool_group.rule_praxes.none? || tool_group.rule_countries.none?
+
+    rule_languages = tool_group.rule_languages.first.languages
+    (rule_languages & languages).any?
```

Comment:
TypeError: no implicit conversion of nil into Array

https://app.rollbar.com/a/Cru/fix/item/mobile-content-api/320

Reply:
```
curl --location --globoff 'https://mobile-content-api-stage.cru.org/resources/suggestions?filter[country]=FR'
```

Reply:
I'll revise this now

Reply:
Take a look please, I think I have solved the problem